private/bufpkg/bufcas/file_set.go

```diff
@@ -0,0 +1,191 @@
+	}
+	return newFileSet(manifest, blobSet), nil
+}
+
+// NewFileSetForBucket returns a new FileSet for the given ReadBucket.
+func NewFileSetForBucket(ctx context.Context, bucket storage.ReadBucket) (FileSet, error) {
+	var fileNodes []FileNode
+	var blobs []Blob
+	if err := storage.WalkReadObjects(
+		ctx,
+		bucket,
+		"",
+		func(readObject storage.ReadObject) error {
+			blob, err := NewBlobForContent(DigestTypeShake256, readObject)
+			if err != nil {
+				return err
+			}
+			fileNode, err := NewFileNode(readObject.Path(), blob.Digest())
+			if err != nil {
+				return err
+			}
+			fileNodes = append(fileNodes, fileNode)
+			blobs = append(blobs, blob)
+			return nil
+		},
+	); err != nil {
+		return nil, err
+	}
+	manifest, err := NewManifest(fileNodes)
+	if err != nil {
+		return nil, err
+	}
+	return newFileSet(
+		manifest,
+		newBlobSet(blobs),
+	), nil
+}
+
+// PutFileSetToBucket writes the FileSet to the given WriteBucket.
+func PutFileSetToBucket(
+	ctx context.Context,
+	fileSet FileSet,
+	bucket storage.WriteBucket,
+) error {
+	for _, fileNode := range fileSet.Manifest().FileNodes() {
+		writeObjectCloser, err := bucket.Put(ctx, fileNode.Path(), storage.PutWithAtomic())
+		if err != nil {
+			return err
+		}
+		blob := fileSet.BlobSet().GetBlob(fileNode.Digest())
+		if _, err := writeObjectCloser.Write(blob.Content()); err != nil {
+			return err
+		}
+		if err := writeObjectCloser.Close(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// FileSetToProtoManifestBlobAndBlobs converts the given FileSet into a proto Blob representing the
+// Manifest, and a set of Blobs representing the Files.
+//
+// TODO: validate the returned proto Blobs.
+func FileSetToProtoManifestBlobAndBlobs(fileSet FileSet) (*storagev1beta1.Blob, []*storagev1beta1.Blob, error) {
+	protoManifestBlob, err := ManifestToProtoBlob(fileSet.Manifest())
+	if err != nil {
+		return nil, nil, err
+	}
+	protoBlobs, err := BlobSetToProtoBlobs(fileSet.BlobSet())
+	if err != nil {
+		return nil, nil, err
+	}
+	return protoManifestBlob, protoBlobs, nil
+}
+
+// ManifestBlobsAndBlobsToFileSet converts the given manifest Blob and set of Blobs representing
+// the Files into a FileSet.
+//
+// Validation is done to ensure the Manifest exactly matches the BlobSet.
+// TODO: validate the input proto Blobs.
+func ProtoManifestBlobsAndBlobsToFileSet(
+	protoManifestBlob *storagev1beta1.Blob,
+	protoBlobs []*storagev1beta1.Blob,
+) (FileSet, error) {
+	manifest, err := ProtoBlobToManifest(protoManifestBlob)
+	if err != nil {
+		return nil, err
+	}
+	blobSet, err := ProtoBlobsToBlobSet(protoBlobs)
+	if err != nil {
+		return nil, err
+	}
+	return NewFileSet(manifest, blobSet)
+}
+
+// *** PRIVATE ****
+
+type fileSet struct {
+	manifest Manifest
+	blobSet  BlobSet
+}
+
+func newFileSet(manifest Manifest, blobSet BlobSet) *fileSet {
+	return &fileSet{
```

Comment:
We generally don't do nil checking for parameters, see other comments. I think I added it for `FileNode` because I was already doing it for path, and path has some weird behavior on empty for `NormalizeAndValidate`. Not a great argument but.